discovery-handlers/onvif/src/discovery_handler.rs

```diff
@@ -0,0 +1,635 @@
+        filter_count == 0
+    } else {
+        filter_count != 0
+    }
+}
+
+async fn apply_filters(
+    discovery_handler_config: &OnvifDiscoveryHandlerConfig,
+    device_service_uris: Vec<String>,
+    onvif_query: &impl OnvifQuery,
+) -> Result<Vec<Device>, anyhow::Error> {
+    let mut result = Vec::new();
+    for device_service_url in device_service_uris.iter() {
+        trace!("apply_filters - device service url {}", &device_service_url);
+        let (ip_address, mac_address) = match onvif_query
+            .get_device_ip_and_mac_address(&device_service_url)
+            .await
+        {
+            Ok(ip_and_mac) => ip_and_mac,
+            Err(e) => {
+                error!("apply_filters - error getting ip and mac address: {}", e);
+                continue;
+            }
+        };
+
+        // Evaluate camera ip address against ip filter if provided
+        let ip_address_as_vec = vec![ip_address.clone()];
+        if execute_filter(
+            discovery_handler_config.ip_addresses.as_ref(),
+            &ip_address_as_vec,
+        ) {
+            continue;
+        }
+
+        // Evaluate camera mac address against mac filter if provided
+        let mac_address_as_vec = vec![mac_address.clone()];
+        if execute_filter(
+            discovery_handler_config.mac_addresses.as_ref(),
+            &mac_address_as_vec,
+        ) {
+            continue;
+        }
+
+        let ip_and_mac_joined = format!("{}-{}", &ip_address, &mac_address);
+
+        // Evaluate camera scopes against scopes filter if provided
+        let device_scopes = match onvif_query.get_device_scopes(&device_service_url).await {
+            Ok(scopes) => scopes,
+            Err(e) => {
+                error!("apply_filters - error getting scopes: {}", e);
+                continue;
+            }
+        };
+        if execute_filter(discovery_handler_config.scopes.as_ref(), &device_scopes) {
+            continue;
+        }
+
+        let mut properties = HashMap::new();
+        properties.insert(
+            ONVIF_DEVICE_SERVICE_URL_LABEL_ID.to_string(),
+            device_service_url.to_string(),
+        );
+        properties.insert(ONVIF_DEVICE_IP_ADDRESS_LABEL_ID.into(), ip_address);
+        properties.insert(ONVIF_DEVICE_MAC_ADDRESS_LABEL_ID.into(), mac_address);
+
+        trace!(
+            "apply_filters - returns DiscoveryResult ip/mac: {:?}, props: {:?}",
+            &ip_and_mac_joined,
+            &properties
+        );
+        result.push(Device {
+            id: ip_and_mac_joined,
+            properties,
+            mounts: Vec::default(),
+            device_specs: Vec::default(),
+        })
+    }
+    Ok(result)
+}
+
+/// This obtains the `OnvifDiscoveryHandlerConfig` from a discovery details map.
+/// It expects the `OnvifDiscoveryHandlerConfig` to be serialized yaml stored in the map as
+/// the String value associated with the key `protocolHandler`.
+fn deserialize_discovery_details(
```

Comment:
can `deserialize_discovery_details` be provided by the util library?

Reply:
fixed in 0c51141